staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go

```diff
@@ -531,52 +531,121 @@ var (
-		if f == metav1.FinalizerOrphan {
+		// validation should make sure the two cases won't be true at the same
+		// time.
+		switch f {
+		case metav1.FinalizerOrphanDependents:
 			shouldOrphan = true
 			hasOrphanFinalizer = true
 			break
+		case metav1.FinalizerDeleteDependents:
+			shouldOrphan = false
+			break
 		}
-		// TODO: update this when we add a finalizer indicating a preference for the other behavior
 	}
+
 	// If an explicit policy was set at deletion time, that overrides both
 	if options != nil && options.OrphanDependents != nil {
 		shouldOrphan = *options.OrphanDependents
 	}
-	if shouldOrphan && !hasOrphanFinalizer {
-		finalizers = append(finalizers, metav1.FinalizerOrphan)
-		return true, finalizers
+	if options != nil && options.PropagationPolicy != nil {
+		switch *options.PropagationPolicy {
+		case metav1.DeletePropagationOrphan:
+			shouldOrphan = true
+		case metav1.DeletePropagationBackground, metav1.DeletePropagationForeground:
+			shouldOrphan = false
+		}
 	}
-	if !shouldOrphan && hasOrphanFinalizer {
-		var newFinalizers []string
-		for _, f := range finalizers {
-			if f == metav1.FinalizerOrphan {
-				continue
-			}
-			newFinalizers = append(newFinalizers, f)
+
+	shouldUpdate = shouldOrphan != hasOrphanFinalizer
+	return shouldUpdate, shouldOrphan
+}
+
+// shouldUpdateFinalizerDeleteDependents returns if the finalizers need to be
+// updated for FinalizerDeleteDependents. In the order of highest to lowest
+// priority, there are three factors affect whether to add/remove the
+// FinalizerDeleteDependents: options, existing finalizers of the object, and
+// e.DeleteStrategy.DefaultGarbageCollectionPolicy.
+func shouldUpdateFinalizerDeletingDependents(e *Store, accessor metav1.Object, options *metav1.DeleteOptions) (shouldUpdate bool, shouldDeleteInForeground bool) {
+	// default to false
+	shouldDeleteInForeground = false
+
+	// If a finalizer is set in the object, it overrides the default
+	hasFinalizerDeleteDependents := false
+	finalizers := accessor.GetFinalizers()
+	for _, f := range finalizers {
+		// validation has made sure the two cases won't be true at the same
+		// time.
+		switch f {
+		case metav1.FinalizerDeleteDependents:
+			shouldDeleteInForeground = true
+			hasFinalizerDeleteDependents = true
+			break
+		case metav1.FinalizerOrphanDependents:
+			shouldDeleteInForeground = false
+			break
 		}
-		return true, newFinalizers
 	}
-	return false, finalizers
+
+	// If an explicit policy was set at deletion time, that overrides both
+	if options != nil && options.OrphanDependents != nil {
+		shouldDeleteInForeground = false
+	}
+	if options != nil && options.PropagationPolicy != nil {
+		switch *options.PropagationPolicy {
+		case metav1.DeletePropagationForeground:
+			shouldDeleteInForeground = true
+		case metav1.DeletePropagationBackground, metav1.DeletePropagationOrphan:
+			shouldDeleteInForeground = false
+		}
+	}
+
+	shouldUpdate = shouldDeleteInForeground != hasFinalizerDeleteDependents
```

Comment:
Renamed the local variable `shouldDeleteInForeground` to `shouldDeleteDependentsInForeground` to be a little more consistent. The finalizer should be named a `FinalizerDeleteDependentsInForeground`, but it's too long to type everywhere.